### PR TITLE
[device_info_plus] add Windows support 

### PR DESCRIPTION
**Pull request by Schwusch**
## Description

Add support for Windows in `package:device_info_plus`. This draft PR is intended for early feedback and clarity on where we stand on Windows support.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [ ] All existing and new tests are passing.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [X] I read and followed the [Flutter Style Guide].
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [X] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [ ] No, this is *not* a breaking change.

### CODE REVIEW

**Comment by jpnurmi** [comment]
Looking good so far. I'll spare the nit-picks for later. :smile: Are there any issues or concerns?

Perhaps we could include the computer name? https://github.com/timsneath/win32/blob/main/example/sysinfo.dart#L72

**Comment by Schwusch** [comment]
I went with the decision to migrate the `device_info_plus` packages to null safety, since the null-safe `package:ffi` API changed significantly in 1.0.0, and we're going to migrate sooner or later anyway.
However, I'm having trouble running the example without `--no-sound-null-safety`, and it's unclear why.
Melos is giving me a hard time on Windows, and since I'm overriding dependencies manually while working, I might have missed  to fully migrate something somewhere.

File: packages/device_info_plus_platform_interface/lib/model/windows_device_info.dart [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    required this.computerName,
+    required this.nbrOfCores,
+    required this.systemMemoryInMegabytes,
+  });
+
+  /// The computer's fully-qualified DNS name, where available.
+  final String computerName;
+
+  /// Number of CPU cores on the local machine
+  final int nbrOfCores;
```
Comment (jpnurmi) [inline packages/device_info_plus_platform_interface/lib/model/windows_device_info.dart:18]: Could it be `numberOfCores`?

File: packages/device_info_plus_windows/CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+## 0.0.1
```
Comment (jpnurmi) [inline packages/device_info_plus_windows/CHANGELOG.md:1]: 0.1.0

File: packages/device_info_plus_windows/README.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# Device Info Plus macOS
```
Comment (jpnurmi) [inline packages/device_info_plus_windows/README.md:1]: Windows

File: packages/device_info_plus_windows/lib/src/device_info_plus_windows_real.dart [hunk 4]
```diff
@@ -0,0 +1,89 @@
+/// The Windows implementation of `package_info_plus`.
+library package_info_plus_windows;
+
+import 'dart:ffi';
+
+import 'package:device_info_plus_platform_interface/device_info_plus_platform_interface.dart';
+import 'package:ffi/ffi.dart';
+import 'package:win32/win32.dart';
+
+/// The Windows implementation of [PackageInfoPlatform].
```
Comment (jpnurmi) [inline packages/device_info_plus_windows/lib/src/device_info_plus_windows_real.dart:10]: DeviceInfoPlatform

File: packages/device_info_plus_windows/lib/src/device_info_plus_windows_real.dart [hunk 5]
```diff
@@ -0,0 +1,89 @@
+/// The Windows implementation of `package_info_plus`.
+library package_info_plus_windows;
```
Comment (jpnurmi) [inline packages/device_info_plus_windows/lib/src/device_info_plus_windows_real.dart:2]: device_info_plus

File: packages/device_info_plus_windows/lib/src/device_info_plus_windows_real.dart [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+import 'dart:ffi';
+
+import 'package:device_info_plus_platform_interface/device_info_plus_platform_interface.dart';
+import 'package:ffi/ffi.dart';
+import 'package:win32/win32.dart';
+
+/// The Windows implementation of [PackageInfoPlatform].
+class DeviceInfoWindows extends DeviceInfoPlatform {
+  /// Returns a map with the following keys:
+  /// appName, packageName, version, buildNumber
```
Comment (jpnurmi) [inline packages/device_info_plus_windows/lib/src/device_info_plus_windows_real.dart:13]: needs updating :)

File: packages/device_info_plus_windows/lib/src/device_info_plus_windows_real.dart [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  @override
+  Future<WindowsDeviceInfo> windowsInfo() {
+    //final info = _FileVersionInfo(Platform.resolvedExecutable);
```
Comment (jpnurmi) [inline packages/device_info_plus_windows/lib/src/device_info_plus_windows_real.dart:16]: -

**Review by jpnurmi — COMMENTED** [review]
Did the migration require any manual changes? I'm wondering if it's not too much trouble, we could get them in first to reduce a bit of noise and focus on one thing.

**Comment by Schwusch** [comment]
> Did the migration require any manual changes? I'm wondering if it's not too much trouble, we could get them in first to reduce a bit of noise and focus on one thing.

I did it by hand, since it's not much code. I'll make an attempt of the migration separately.

**Comment by Schwusch** [comment]
Moved the null safety migration to #136. Let's do that first.

File: packages/device_info_plus_windows/lib/src/device_info_plus_windows_real.dart [hunk 8]
```diff
@@ -0,0 +1,87 @@
+/// The Windows implementation of `device_info_plus`.
+library package_info_plus_windows;
```
Comment (jpnurmi) [inline packages/device_info_plus_windows/lib/src/device_info_plus_windows_real.dart:2]: device

File: packages/device_info_plus_windows/README.md (hunk 3 above)
Comment (jpnurmi) [reply packages/device_info_plus_windows/README.md:1]: don't forget this :)

**Review by jpnurmi — COMMENTED** [review]
LGTM :+1:

File: packages/device_info_plus_windows/lib/src/device_info_plus_windows_real.dart (hunk 8 above)
Comment (Schwusch) [reply packages/device_info_plus_windows/lib/src/device_info_plus_windows_real.dart:2]: Right, good catch

**Review by jpnurmi — APPROVED** [review]
Sorry, I forgot one thing... Can we update platform support checkmarks in the README files & the website while at it?

- /README.md
- /packages/device_info_plus/README.md
- /website/plugins.js

**Comment by mhadaily** [comment]
This is amazing, @jpnurmi @Schwusch just make sure when it's merged, the website and readme file are up to date too.

**Review by jpnurmi — APPROVED** [review]
Awesomeness! :tada:

**Comment by Schwusch** [comment]
> This is amazing, @jpnurmi @Schwusch just make sure when it's merged, the website and readme file are up to date too.

Thanks, I'm unsure to do that though. @jpnurmi halp!

**Comment by jpnurmi** [comment]
> Thanks, I'm unsure how to do that though. @jpnurmi halp!

Sorry I was busy tweeting! :D Anyway, no worries, the website gets automatically built and deployed.

**Pull request merged**